pynws/forecast.py

```diff
@@ -0,0 +1,81 @@
+"""Forecast class"""
+import re
+from datetime import datetime, timedelta, timezone
+
+
+ISO8601_PERIOD_REGEX = re.compile(
+    r"^P"
+    r"((?P<weeks>\d+)W)?"
+    r"((?P<days>\d+)D)?"
+    r"((?:T)"
+    r"((?P<hours>\d+)H)?"
+    r"((?P<minutes>\d+)M)?"
+    r"((?P<seconds>\d+)S)?"
+    r")?$"
+)
+
+
+class Forecast:  # pylint: disable=too-few-public-methods
+    """Class to retrieve forecast layer values for a point in time."""
+
+    def __init__(self, properties):
+        if not isinstance(properties, dict):
+            raise TypeError(f"{properties!r} is not a dictionary")
+
+        self.update_time = datetime.fromisoformat(properties["updateTime"])
+        self.layers = layers = {}
+
+        for prop_name, prop_value in properties.items():
+            if not isinstance(prop_value, dict) or "values" not in prop_value:
+                continue
+
+            layer_values = []
+
+            for value in prop_value["values"]:
+                isodatetime, duration_str = value["validTime"].split("/")
+                start_time = datetime.fromisoformat(isodatetime)
+                end_time = start_time + Forecast._parse_duration(duration_str)
+                layer_values.append((start_time, end_time, value["value"]))
+
+            units = prop_value.get("uom")
+            units = units.split(":")[-1] if units else None
+            layers[prop_name] = (layer_values, units)
+
+    @staticmethod
+    def _parse_duration(duration_str):
+        match = ISO8601_PERIOD_REGEX.match(duration_str)
+        groups = match.groupdict()
+
+        for key, val in groups.items():
+            groups[key] = int(val or "0")
+
+        return timedelta(
+            weeks=groups["weeks"],
+            days=groups["days"],
+            hours=groups["hours"],
+            minutes=groups["minutes"],
+            seconds=groups["seconds"],
+        )
+
+    @property
+    def last_update(self):
+        """When the forecast was last updated"""
+        return self.update_time
+
+    def get_forecast_for_time(self, when):
```

Comment:
Does it make sense to allow `when` to be an Iterable too, like a list?  A future implementation could have an additional method to automatically generate an hourly forecast for `x` days or hours.

Reply:
> 1.  Are you thinking that 'when' could contain a list of datetime values?


Yes.

> 2. I like the idea of generating _N_ forecast values for hours or days. That could replace the original gridpoint hourly and daily forecasts which could eliminate a lot of code.

I think it should be a parallel option.  The daily endpoint in particular has a nice human readable forecast.

Reply:
> > 1. Are you thinking that 'when' could contain a list of datetime values?
> 
> Yes.

It would probably be better to have a separate method for that because the return value would have a different shape (like the output from the next item)

> > 2. I like the idea of generating _N_ forecast values for hours or days. That could replace the original gridpoint hourly and daily forecasts which could eliminate a lot of code.
> 
> I think it should be a parallel option. The daily endpoint in particular has a nice human readable forecast.

Fair point.

Reply:
> Does it make sense to allow `when` to be an Iterable too, like a list? A future implementation could have an additional method to automatically generate an hourly forecast for `x` days or hours.

These have both been added (for _X_ hours--not days)